File: packages/client/src/pipeline/transport.ts
import {
  CORRELATION_ID_HEADER,
  NetworkError,
  TimeoutError,
  errorFromEnvelope,
  type ErrorEnvelope,
  type HttpMethod,
  type ResponseEnvelope,
} from "@ecorpin/core";

export interface TransportRequest {
  url: string;
  method: HttpMethod;
  authHeader: { name: "Authorization"; value: string };
  body?: unknown;
  timeoutMs: number;
  correlationId: string;
  service: string;
  resource: string;
  action: string;
}

/**
 * The single seam where an actual HTTP request happens. Everything above
 * this layer (proxy engine, retry, auth) works in terms of
 * service/resource/action; this is the one place that knows about `fetch`,
 * headers, and JSON bodies — kept intentionally small so a future
 * transport swap (HTTP/2, gRPC — architecture doc §20) only touches this
 * file.
 */
export async function performRequest(request: TransportRequest): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), request.timeoutMs);

  const headers: Record<string, string> = {
    [request.authHeader.name]: request.authHeader.value,
    [CORRELATION_ID_HEADER]: request.correlationId,
    Accept: "application/json",
  };
  if (request.body !== undefined) {
    headers["Content-Type"] = "application/json";
  }

  let response: Response;
  try {
    response = await fetch(request.url, {
      method: request.method,
      headers,
      body: request.body !== undefined ? JSON.stringify(request.body) : undefined,
      signal: controller.signal,
    });
  } catch (err) {
    if (controller.signal.aborted) {
      throw new TimeoutError(`Request timed out after ${request.timeoutMs}ms.`, {
        service: request.service,
        resource: request.resource,
        action: request.action,
        correlationId: request.correlationId,
      });
    }
    throw new NetworkError(err instanceof Error ? err.message : "Network request failed.", {
      service: request.service,
      resource: request.resource,
      action: request.action,
      correlationId: request.correlationId,
      cause: err,
    });
  } finally {
    clearTimeout(timer);
  }

  const json = await response.json().catch(() => undefined);

  if (!response.ok) {
    const envelope: ErrorEnvelope =
      json && typeof json === "object" && "error" in (json as Record<string, unknown>)
        ? (json as ErrorEnvelope)
        : { error: { code: "ECORPIN_INTERNAL_ERROR", message: `HTTP ${response.status}` } };
    throw errorFromEnvelope(envelope, {
      service: request.service,
      resource: request.resource,
      action: request.action,
      correlationId: request.correlationId,
    });
  }

  return (json as ResponseEnvelope | undefined)?.data;
}

File: packages/client-py/src/ecorpin_client/transport.py
"""
The single seam where an actual HTTP request happens. Everything above this
layer (SDK namespaces, retry, auth) works in terms of service/resource/action;
this is the one place that knows about `requests`, headers, and JSON bodies —
kept intentionally small so a future transport swap only touches this file.
Mirrors @ecorpin/client's pipeline/transport.ts.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import requests

from .constants import CORRELATION_ID_HEADER
from .errors import NetworkError, TimeoutError, error_from_envelope


@dataclass
class TransportRequest:
    url: str
    method: str
    auth_header: tuple[str, str]
    body: Any
    timeout_ms: int
    correlation_id: str
    service: str
    resource: str
    action: str


def perform_request(request: TransportRequest, session: requests.Session) -> Any:
    headers = {
        request.auth_header[0]: request.auth_header[1],
        CORRELATION_ID_HEADER: request.correlation_id,
        "Accept": "application/json",
    }
    json_body = None
    if request.body is not None:
        headers["Content-Type"] = "application/json"
        json_body = request.body

    try:
        response = session.request(
            request.method,
            request.url,
            headers=headers,
            json=json_body,
            timeout=request.timeout_ms / 1000,
        )
    except requests.exceptions.Timeout as err:
        raise TimeoutError(
            f"Request timed out after {request.timeout_ms}ms.",
            service=request.service,
            resource=request.resource,
            action=request.action,
            correlation_id=request.correlation_id,
            cause=err,
        ) from err
    except requests.exceptions.RequestException as err:
        raise NetworkError(
            str(err) or "Network request failed.",
            service=request.service,
            resource=request.resource,
            action=request.action,
            correlation_id=request.correlation_id,
            cause=err,
        ) from err

    try:
        payload = response.json()
    except ValueError:
        payload = None

    if not response.ok:
        envelope = (
            payload
            if isinstance(payload, dict) and "error" in payload
            else {"error": {"code": "ECORPIN_INTERNAL_ERROR", "message": f"HTTP {response.status_code}"}}
        )
        raise error_from_envelope(
            envelope,
            service=request.service,
            resource=request.resource,
            action=request.action,
            correlation_id=request.correlation_id,
        )

    return (payload or {}).get("data") if isinstance(payload, dict) else None
